### PR TITLE
Add VACF implementation

### PR DESCRIPTION
**Pull request by xhgchen**
This Pull Request will implement two methods for calculating a velocity autocorrelation function (VACF) from an `AtomGroup`. The default computation is a fast Fourier transform-based (FFT) Fast Correlation Algorithm (FCA) powered by the `tidynamics` package with $O(nlogn)$ running time. An alternative simple "windowed" calculation provides $O(n^2)$ running time.

Changes made in this Pull Request:

 - Add items to `__init__()`
 - Write `_prepare()` and `_parse_dim_type()`
 - Write `_single_frame()`
 - Add MDAnalysisTests to access AMBER test data files with velocities
 - Provide exception handling
 - Implement VACF computation via FFT
 - Implement simple "windowed" algorithm for VACF calculation
 - Add tests for method signatures, VACF FFT implementation, and VACF simple "windowed" implementation

### CODE REVIEW

File: transport_analysis/analysis/velocityautocorr.py
Comment (hmacdope) [inline transport_analysis/analysis/velocityautocorr.py:27]: You could probably make it so that `yx` is ok, I think it would still make logical sense in that the dims don't have an order except by convention, would still follow the `xy` code-path

File: transport_analysis/analysis/velocityautocorr.py
Comment (hmacdope) [inline transport_analysis/analysis/velocityautocorr.py:56]: Docstring? / what is `times` going to do?

File: transport_analysis/analysis/velocityautocorr.py
Comment (hmacdope) [inline transport_analysis/analysis/velocityautocorr.py:65]: Lets consider making a decorator that returns `u.atoms` if the input is a universe, that way you can accept both.

File: transport_analysis/analysis/velocityautocorr.py
Comment (hmacdope) [inline transport_analysis/analysis/velocityautocorr.py:118]: Have a look at how this is done in `MSD` class. :)

**Review by hmacdope — CHANGES_REQUESTED** [review]
Looking great already, a few hints to help you on your way.

File: transport_analysis/analysis/velocityautocorr.py
Comment (orionarcher) [inline transport_analysis/analysis/velocityautocorr.py:101]: Maybe instantiate these as empty version of the data type you intend them to be?

**Review by orionarcher — COMMENTED** [review]
Looking good!

We didn't discuss this previously, but I'd advocate for setting up a linter for your code, [black](https://black.readthedocs.io/en/stable/) is a popular one. It saves mental effort and enforces consistent coding style, which is nice. It's also possible to use a tool like [pre-commit](https://pre-commit.com/) which can automatically reformats when you commit. This is another item in the "I wish I had done this sooner when making solvation-analysis" bucket.

It's not in the cookiecutter by default, but @hmacdope or I can advise on setup, assuming you and Hugo think it's a good idea :)

File: setup.py
Comment (lilyminium) [inline setup.py:29]: Yeah unfortunately `setup` doesn't support lists of strings. The only workaround is to make the whole thing one string, i.e. `'Xu Hong Chen, Hugo MacDermott-Opeskin, Orion Cohen'`, but it's probably easier to just keep `AUTHORS.md` up-to-date.

File: transport_analysis/analysis/velocityautocorr.py
Comment (xhgchen) [reply transport_analysis/analysis/velocityautocorr.py:27]: So let it accept all possible permutations? Sure, that would be good.

File: transport_analysis/analysis/velocityautocorr.py
Comment (xhgchen) [reply transport_analysis/analysis/velocityautocorr.py:56]: `times` likely will not be used explicitly in this class. Would it be best to remove it from the docstring then?

Comment (hmacdope) [reply transport_analysis/analysis/velocityautocorr.py:56]: Probably.

File: transport_analysis/analysis/velocityautocorr.py
Comment (xhgchen) [reply transport_analysis/analysis/velocityautocorr.py:65]: What do you mean by decorator? I was following the feedback in my proposal about avoiding having a `Universe` with a selection string and only using an `AtomGroup` instead.

File: transport_analysis/analysis/velocityautocorr.py
Comment (xhgchen) [reply transport_analysis/analysis/velocityautocorr.py:101]: I do that in `_prepare()`, just like in the MSD module. Would it be better to do it here? Have not tried it myself

File: transport_analysis/analysis/velocityautocorr.py
Comment (xhgchen) [reply transport_analysis/analysis/velocityautocorr.py:118]: The code here is the same as in the `MSD` class, so it should work, right? Iirc it worked for VACF back when I tested it for the proposal and a week or two ago. Then I would just add support for more permutations in `keys`?

Comment (hmacdope) [reply transport_analysis/analysis/velocityautocorr.py:118]: Yeah, but on reevaluation, lets hold off on all permutation support for now, this should do until we establish the functionality we want.

File: transport_analysis/analysis/velocityautocorr.py
Comment (hmacdope) [reply transport_analysis/analysis/velocityautocorr.py:65]: Just a thought I have been throwing around, would look like 

```python
@universe_to_atomgroup # decorator
def my_func_that_takes_u_or_ag(input):
      # decorator converts input universe to `atomgroup` if it isnt one already
      do_something_with_atomgroup(input)
```

But tbh this is just a random nice to have, and should probably go in MDA-core, so ignore me for now hahahah.

File: transport_analysis/analysis/velocityautocorr.py
Comment (orionarcher) [reply transport_analysis/analysis/velocityautocorr.py]: Hmm, IMO it's nicer in terms of readability to have it only in one place, your call on where it goes. Defining empty versions is also nice because the types are clear.

File: transport_analysis/analysis/velocityautocorr.py
Comment (hmacdope) [inline transport_analysis/analysis/velocityautocorr.py:61]: Does the `AtomGroup` type hint not work here once you have imported it?

File: transport_analysis/analysis/velocityautocorr.py
Comment (hmacdope) [inline transport_analysis/analysis/velocityautocorr.py:86]: Agree with @orionarcher, lets just allocate the arrays here rather than in _prepare.

File: transport_analysis/tests/analysis/test_velocityautocorr.py
Comment (hmacdope) [inline transport_analysis/tests/analysis/test_velocityautocorr.py:53]: Can you also write a test that checks `characteristic_poly` vs `tidynamics.acf`

**Review by hmacdope — CHANGES_REQUESTED** [review]
Looking great! Just a few small comments.

**Review by hmacdope — CHANGES_REQUESTED** [review]
Looking great! Just a few small comments.

File: transport_analysis/analysis/velocityautocorr.py
Comment (xhgchen) [reply transport_analysis/analysis/velocityautocorr.py:86]: So I tried this and it seems like it does not work because `n_frames` is not accessible in `__init__()`

Comment (lilyminium) [reply transport_analysis/analysis/velocityautocorr.py:86]: If you're very keen on having empty arrays there, one option is to set up `np.empty((0, ...), dtype=float)`. That could be confusing for users though.

Comment (xhgchen) [reply transport_analysis/analysis/velocityautocorr.py:87]: If that is the only alternative, I feel the current setup would be a little nicer. @hmacdope @orionarcher What are your thoughts?

**Review by hmacdope — CHANGES_REQUESTED** [review]
Few small changes but looking great. @orionarcher would be great to get another pair of eyes here.

File: transport_analysis/analysis/velocityautocorr.py
Comment (orionarcher) [inline transport_analysis/analysis/velocityautocorr.py:116]: It's mentioned above to include any ordering of `xyz`, which I think would be nice. A good way to implement that would be to convert the `dim_type` string to a set and then use sets as the keys here. That would give you flexibility without enumerating every option. e.g.

```python
dim_type = set(self.dim_type.split())

        keys = {
            {"x"}: [0],
            {"y"}: [1],
            {"z"}: [2],
            {"x", "y"}: [0, 1],
            {"x", "z"}: [0, 2],
            {"y", "z"}: [1, 2],
            {"x", "y", "z"}: [0, 1, 2],
        }

self._dim = keys[dim_type]

```

File: transport_analysis/analysis/velocityautocorr.py
Comment (orionarcher) [reply transport_analysis/analysis/velocityautocorr.py:87]: Current setup meaning allocating them in _prepare instead of in __init__? I'd be happy with that option too. I like what you have here once the redundant `None` assignments are removed.

File: transport_analysis/analysis/velocityautocorr.py
Comment (orionarcher) [inline transport_analysis/analysis/velocityautocorr.py:174]: I fear that because you are using a triple quote the white space will be rendered in the error message and make it less readable. Could you check that? If so, you could replace these with a set of quotes in parentheses. e.g.

```python
long_quote = (
    " text text text",
    " text text text",
    " text text text",
    " text text text"
)
```

I think thats the correct syntax, you may not even need the commas

File: transport_analysis/analysis/velocityautocorr.py
Comment (orionarcher) [reply transport_analysis/analysis/velocityautocorr.py:65]: If this were my code I'd only accept AtomGroup and not Universe. AtomGroup is significantly more flexible than Universe and feels conceptually like the right thing here as we are ultimately interested in the VACF of the particles. It's not much work to support Universe, but it's even less work to turn a universe into an atom group (`u.atoms`). Ultimately a personal preference.

**Review by orionarcher — CHANGES_REQUESTED** [review]
Looking good! Just a few comments and suggestions.

File: transport_analysis/analysis/velocityautocorr.py
Comment (xhgchen) [reply transport_analysis/analysis/velocityautocorr.py:116]: This looks great! Would you be all right with me making this change in a separate PR? I'd like to get this one merged first to fix the CI (the check via Black is failing atm).

File: transport_analysis/analysis/velocityautocorr.py
Comment (xhgchen) [reply transport_analysis/analysis/velocityautocorr.py:87]: Ah, you're right! They are redundant!

File: transport_analysis/analysis/velocityautocorr.py
Comment (xhgchen) [reply transport_analysis/analysis/velocityautocorr.py:174]: So I've checked it on my machine (Fedora Linux 38) and it shows up fine in both the terminal and Jupyter Lab. Feel free to check it on yours as well and let me know if you see anything strange!

File: transport_analysis/tests/analysis/test_velocityautocorr.py
Comment (xhgchen) [reply transport_analysis/tests/analysis/test_velocityautocorr.py:86]: Could you elaborate? Does that mean I should just remove the "try" and "except" part and only have the rest of the test?

File: transport_analysis/analysis/velocityautocorr.py
Comment (xhgchen) [reply transport_analysis/analysis/velocityautocorr.py:61]: I noted this on Discord already but I just wanted to add it here so that the discussion is a bit more public. I tried to use mypy to check whether the `AtomGroup` type hint works, but I ran into some issues:
```
error: Need type annotation for "LONG_VERSION_PY" (hint: "LONG_VERSION_PY: Dict[<type>, <type>] = ...")  [var-annotated]
velocityautocorr.py:10: error: Cannot find implementation or library stub for module named "MDAnalysis.analysis.base"  [import]
velocityautocorr.py:10: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
velocityautocorr.py:154: error: Skipping analyzing "tidynamics": module is installed, but missing library stubs or py.typed marker  [import]
Found 6 errors in 2 files (checked 1 source file)
```
I looked this up for a while but I got stuck. Will be leaving it for the time being to work on the other items, but will revisit it.

File: transport_analysis/analysis/velocityautocorr.py
Comment (orionarcher) [reply transport_analysis/analysis/velocityautocorr.py:175]: Got it, good to know. I am curious if there is any PEP guidance on triple quotes vs parenthetical single quotes, does anyone else have an opinion here?

File: transport_analysis/analysis/velocityautocorr.py
Comment (orionarcher) [reply transport_analysis/analysis/velocityautocorr.py:117]: Yep no problem. Maybe raise an issue and resolve this comment?

File: transport_analysis/analysis/velocityautocorr.py
Comment (xhgchen) [reply transport_analysis/analysis/velocityautocorr.py:39]: Done! I only need to cite the tidynamics package for now, right?

File: transport_analysis/analysis/velocityautocorr.py
Comment (orionarcher) [reply transport_analysis/analysis/velocityautocorr.py:69]: tbh mypy is kinda cursed. Not sure there is anything much better, but be aware of the `# type: ignore` comment if you ever feel like you are wasting too much time with it.

Comment (xhgchen) [reply transport_analysis/analysis/velocityautocorr.py:69]: You're right - I revisited it today and it has been too much trouble for something so minor. I went through a bunch of documentation, discussions, and even ChatGPT but had no luck. None of the errors mention the `AtomGroup` part and I wrote that part based on the cookiecutter template (only taking out the `Universe` part because it was unnecessary), so I think it is probably okay. I also ran it with the `# type: ignore` comment you mentioned. I think, at this point, I have done what I can.

Comment (hmacdope) [reply transport_analysis/analysis/velocityautocorr.py:69]: Sounds good.

File: transport_analysis/tests/analysis/test_velocityautocorr.py
Comment (hmacdope) [reply transport_analysis/tests/analysis/test_velocityautocorr.py:53]: Done thanks

File: transport_analysis/tests/analysis/test_velocityautocorr.py
Comment (hmacdope) [reply transport_analysis/tests/analysis/test_velocityautocorr.py:86]: Yes as in you shouldn't be able to use the package without tidynamics tbh. I think we can be opinionated and insist. We didn't for MSD in MDAnalysis core because of the need to keep our dep list as short as possible

File: transport_analysis/analysis/velocityautocorr.py
Comment (hmacdope) [reply transport_analysis/analysis/velocityautocorr.py:119]: Done in #13

File: transport_analysis/analysis/velocityautocorr.py
Comment (hmacdope) [reply transport_analysis/analysis/velocityautocorr.py:177]: Should be fine with triple quotes but see my  opinion on forcing tidynamics to be installed.

File: transport_analysis/tests/analysis/test_velocityautocorr.py
Comment (xhgchen) [reply transport_analysis/tests/analysis/test_velocityautocorr.py:86]: So that would mean removing the exception and the corresponding test, along with the skip if no tidynamics parts in the tests, right? I've already added tidynamics in `devtools/conda-envs/test_env.yaml` and `pyproject.toml`. Would I add it to the `install_requires=` part of `setup.py` as well? (These are all from https://mdakits.mdanalysis.org/makingakit.html) Any other places?

**Pull request merged**
